### DISCUSSION TITLE
Add device classes to `binary_sensor` for CO and CO2

### DISCUSSION THREAD

**Discussion by epenet** [home-assistant/architecture#690] [Entity Models] [answered]
As I was working on https://github.com/home-assistant/core/pull/61665, I found that device classes `CO` and `CO2` were available on `Sensor` platform, but not on `BinarySensor` platform.

`CO` and `CO2` were originally discussed for `Sensor` platform in https://github.com/home-assistant/architecture/discussions/520

I propose that they get added to `BinarySensor` platform.

**Comment by farmio** [comment]
Just out of interest: what does a CO2 BinarySensor show?

**Reply by epenet** [reply]
It is present in `homekit` code. I assume it is above/below a threshold.
If we do not want it, then should it be considered a bug in `homekit` code?

**Reply by Jc2k** [reply]
From the HomeKit spec:

> This characteristic indicates if a sensor detects abnormal levels of Carbon Dioxide. Value should revert to 0 after the Carbon Dioxide levels drop to normal levels.

**Reply by thomasloven** [reply]
Wouldn't a more generic class be better then. `LEVEL_WARNING` or something? `ALARM` even?

**Reply by Jc2k** [reply]
Specific device classes allow the core to make intelligent helpful choices. In this case, it would allow the binary sensor to be exported as a `CARBON_DIOXIDE_DETECTED` characteristic in HomeKit/Siri automatically without user intervention. 

(And FWIW, there isn't a generic alarm or level warning in HomeKit/Siri)

**Reply by bdraco** [reply]
This makes sense to me. Carbon monoxide detectors are present in many modern homes, and a detected state is going to happen at least once during the lifetime of the device

**Reply by emontnemery** [reply]
I'm still a bit curious about this. Googling for "CO2 alarm" only gets me results from people who confuse CO2 with CO, are there any binary CO2 detectors on the market and if there are are any integrated with Home Assistant?

CO2 sensors are used to monitor air quality, for example https://www.se.com/ww/en/product/SED-CO2-G-5045/wireless-co2-sensor-with-room-temperature-and-humidity-for-se8000-series-zigbee-communication-protocol/
Maybe there are some which have a binary output to for example increase the ventilation?
Or maybe we want to allow the user to make a template binary CO2 sensor which triggers at some CO2 threshold?

**Reply by Jc2k** [reply]
This sort of talks about how this works for a homekit device.

https://www.thinka.eu/faq/how-do-i-configure-my-carbon-dioxide-co2-sensor

**Reply by Jc2k** [reply]
And looking at the HAP spec more closely, a co2 sensor MUST have a binary alert for when dangerous levels are detected. The actual value itself is optional.

**Reply by emontnemery** [reply]
That link explains that it's possible to map a KNX group address to a homekit `CARBON_DIOXIDE_DETECTED` characteristic I think.
This KNX CO2 sensor, https://partner.gira.com/en/produkte/gebaeudetechnik/klima-heizungssteuerung/knx-co2sensor.html allows setting up "limiting value" connected to the CO2 values, which could then presumably be configured as a homekit `CARBON_DIOXIDE_DETECTED` characteristic.

In our case it would probably mean configuring a template binary sensor?

**Reply by Jc2k** [reply]
That's definitely an option that should be supported, but to be clear we are all on the same page: according to the HAP spec every homekit CO and CO2 sensor will have a Boolean "detected" characteristic, and may not have a integer (ppm) level characteristic.

**Comment by epenet** [comment]
Draft implementation PR: https://github.com/home-assistant/core/pull/61962
Draft documentation PR: https://github.com/home-assistant/home-assistant.io/pull/20721
Draft frontend PR: https://github.com/home-assistant/frontend/pull/10935

**Comment by MartinHjelmare** [comment]
What integrations can leverage these device classes?

**Reply by epenet** [reply]
I only bumped into it on `homekit`, but it may be used elsewhere that I didn't see.
My guess is that any component that has `CO` or `CO2` sensor capabilities could be extended to have a corresponding `binary_sensor` (either with a fixed or a configurable threshhold)
I'd also see it being used in `template` entities.

**Reply by MartinHjelmare** [reply]
We need some more actual cases of integrations that will use the device class before we add it. 1-2 integrations is too few.

**Reply by epenet** [reply]
With the DEVICE_CLASS constants being deprecated I am not sure what to do about homekit then. Create a local constant in `homekit` or something else?

**Reply by MartinHjelmare** [reply]
What is the problem?

**Reply by Jc2k** [reply]
If I understand correctly, the device class sort of de facto exists already, it somehow got added to homekit. So if it isn't approved retrospectively we'll need a breaking change to get rid of it before the Enum work can conclude?

I guess the other question is if it's ok to have custom local device classes in homekit to map entities to the right homekit types? (I think this "class" currently is used through entity customisations? I haven't looked in a while so might be wrong there).

Every CO and CO2 sensor integration technically needs a binary_sensor so that it can be shown in home on iOS (even if a template), as the spec says the binary version is the required value. So I assume that this might affect every existing co/co2 sensors ability to be exported to homekit on iOS, if you count that way we have enough integrations 😅

**Reply by bdraco** [reply]
I think any integration that has a smart CO detector could use it.  

That's at least `tuya` (various models, mostly x-sense branded), `nest` (protect),`honeywell` (various models),  first alert via `homekit_controller`,  `ring` (Ring smoke).

I have one I'm pulling data off via `rest` as a `binary_sensor` which could use it as well.  

`esphome` also supports a few CO2 sensors https://esphome.io/components/sensor/mhz19.html https://esphome.io/components/sensor/scd30.html https://esphome.io/components/sensor/scd4x.html

**Reply by bdraco** [reply]
Also it looks like my `elkm1` has the ability to setup a binary_sensor zone as `co` detected

**Reply by emontnemery** [reply]
> Every CO and CO2 sensor integration technically needs a binary_sensor so that it can be shown in home on iOS (even if a template), as the spec says the binary version is the required value.

We can't burden the integrations or the user with limitations in homekit by making it the integration's or user's responsibility  to ensure each CO or CO2 sensor is matched with a corresponding binary sensor though, this probably needs to be handled in the homekit integration itself by binarizing CO or CO2 sensor data if no binary sensor is available in the device?

**Reply by bdraco** [reply]
Ignoring the HomeKit case for a moment, Home CO detectors are usually binary sensors since they have two states:

0: no CO detected - everything is ok 
1: CO detected - risk of CO poisoning - alarm is probably going off

CO2 has much lower risks since you need concentrations of 10% of higher before risking death. Usually this is only going to be some type of binary alarm for a commercial setting, maybe some place that frequently works dry ice such as a vaccination center.

**Reply by emontnemery** [reply]
Ok, so which integrations providing binary sensor entities would benefit from adding binary sensor device classes?

**Reply by epenet** [reply]
With covid-19, I think CO2 binary sensors would be used to trigger "this room needs airing" rather than "risk of death"

**Reply by bdraco** [reply]
Doing a quick product search I find a multitude of CO alarm products that are binary sensors (alarm is on or off, risk of death). 

I couldn't find any CO2 alarm products intended for a Home market. Everything I found measured CO2 in PPM

I support a CO binary sensor device class.

I don't see a justification for a CO2 one though as there doesn't seem to be a home market for these.

**Comment by emontnemery** [comment, marked as answer]
Adding the comment from @bdraco as a concrete proposal: https://github.com/home-assistant/architecture/discussions/690#discussioncomment-1851458, with integrations supporting CO binary sensors listed here: https://github.com/home-assistant/architecture/discussions/690#discussioncomment-1849938

Let's add a CO binary sensor device class.
There's no a justification for a CO2 binary sensor device class at this point as there doesn't seem to be a home market for these.

**Reply by MartinHjelmare** [reply]
Sounds good!

**Reply by epenet** [reply]
I have updated the three PRs to only include carbon monoxide.
